### PR TITLE
Fix ABA issue in LocklessTaskQueue

### PR DESCRIPTION
**Pull request by jkammerl**
See also https://github.com/resonance-audio/resonance-audio/pull/32:

### CODE REVIEW

**Review by anokta — APPROVED** [review]

**Comment by tristanjl** [comment]
Looks like the linux compilers don't like uint32/uint64 - maybe swap for uint32_t and uint64_t.
Otherwise, at first glance, that does seem better - swapping everything to indices and increasing the tag only on return from pop does seem much safer. Gets around worrying about masked writes to the list heads.
I should be able to run some tests in the next day or two.

File: resonance_audio/utils/lockless_task_queue.cc [hunk 1]
```diff
@@ -16,76 +16,114 @@ limitations under the License.
...
+#include <limits>
+
 #include "base/logging.h"
 
 namespace vraudio {
 
+namespace {
+
+// Reserved index representing an invalid list index.
+constexpr uint64 kInvalidIndex = std::numeric_limits<uint32>::max();
```
Comment (tristanjl) [inline resonance_audio/utils/lockless_task_queue.cc]: uint64 -> uint64_t and uint32 -> uint32_t

File: resonance_audio/utils/lockless_task_queue.cc [hunk 2]
```diff
@@ -16,76 +16,114 @@ limitations under the License.
...
+
+// Maximum number of producers.
+constexpr uint64 kMaxProducers = kInvalidIndex - 1;
```
Comment (tristanjl) [inline resonance_audio/utils/lockless_task_queue.cc]: uint64 -> uint64_t

File: resonance_audio/utils/lockless_task_queue.cc [hunk 3]
```diff
@@ -16,76 +16,114 @@ limitations under the License.
...
+  const TagAndIndex old_task_list_head_idx =
+      task_list_head_idx_.exchange(old_flag_with_invalid_index);
+  ProcessTaskList(old_task_list_head_idx, false /*execute_tasks*/);
+}
+
+LocklessTaskQueue::TagAndIndex LocklessTaskQueue::IncreaseTag(
+    TagAndIndex tag_and_index) {
+  // The most significant 32 bits a reserved for tagging. Overflows are
+  // acceptable.
+  return tag_and_index + (static_cast<uint64>(1) << 32);
```
Comment (tristanjl) [inline resonance_audio/utils/lockless_task_queue.cc]: uint64 -> uint64_t

File: resonance_audio/utils/lockless_task_queue.cc [hunk 4]
```diff
@@ -16,76 +16,114 @@ limitations under the License.
...
+}
+
+LocklessTaskQueue::TagAndIndex LocklessTaskQueue::GetIndex(
+    TagAndIndex tag_and_index) {
+  // The least significant 32 bits a reserved for the index.
+  return tag_and_index & std::numeric_limits<uint32>::max();
```
Comment (tristanjl) [inline resonance_audio/utils/lockless_task_queue.cc]: uint32 -> uint32_t

File: resonance_audio/utils/lockless_task_queue.h [hunk 5]
```diff
@@ -17,16 +17,15 @@ limitations under the License.
...
+#include <cstdint>
 #include <functional>
 #include <vector>
 
 namespace vraudio {
 
-// Lock-less task queue which is thread safe for cocurrent task producers and a
-// single task executor thread.
+// Lock-less task queue which is thread safe for concurrent task producers and
+// concurrent task consumers.
```
Comment (tristanjl) [inline resonance_audio/utils/lockless_task_queue.h]: temp_tasks_ is still being shared by consumers, so this second part isn't currently true.

**Review by tristanjl — COMMENTED** [review]
Did some testing and the changes seem good with the minor adjustments.

**Pull request merged**